Allow jt to create flamegraphs for CRuby

rwstauner opened the pull request:
This adds a wrapper script and installs some gems
so that `jt profile ...` will produce a flamegraph
even with `RUBY_BIN` set to CRuby.

Also allow specifying a suffix for the svg files
to make it easier to remember what they represent.

example usage:

```
JT_PROFILE_SUFFIX=cruby RUBY_BIN=`which ruby` jt profile -e 'puts '`
```

eregon reviewed (commented):
I think it's a good idea in general but there is quite a few specifics we need to discuss

eregon commented on tool/jt.rb:
```diff
@@ -1114,6 +1114,17 @@ def gem(*args)
     ruby '-S', 'gem', *args
   end
 
+  def ensure_gem_installed(gem_name, version = '', lib = gem_name)
+    # Use `gem_name.upcase` to match what rubygems.rb does.
+    env = {"GEM_REQUIREMENT_#{gem_name.upcase}" => "= #{version}"}
+
+    unless run_ruby(env, "-r#{lib}", '-e1', {:continue_on_failure => true, [:err, :out] => :close})
+      run_ruby({}, '-S', 'gem', 'install', gem_name, '-v', version)
+    end
+
+    env
+  end
```
This feels too brittle. I would much rather just let the program fail if the necessary gem is not installed, and let the user install it.
FWIW, there is `Gem.install` so it's possible to install gems without needed extra subprocesses, e.g., in tool/profile_mri.rb

rwstauner commented:
> I think we should just have that behavior on by default, e.g, append -cruby if RUBY_ENGINE == "ruby", since anyway the filenames include the time and so are not predictable anyway.

I had originally put `@ruby_name` as the suffix, but then found myself wanting even more info (for example, i rewrote `Regexp.match?` to see how the flame changed, then restored that and rewrote a different method and made another one...)

rwstauner commented on tool/jt.rb:
```diff
@@ -2094,14 +2109,28 @@ def profile(*args)
+                 else
+                   repo = flamegraph_repo
+                   env = env.merge(
+                     ensure_gem_installed('ruby-prof', '0.18.0'),
+                     ensure_gem_installed('ruby-prof-flamegraph', '0.3.0'),
```
this is what i found searching for ruby flamegraphs, so i'm happy to try out another way.
if it didn't need specific gem versions it would make things a lot simpler.
i'll give this a try, thanks!

eregon commented:
> I had originally put `@ruby_name` as the suffix, but then found myself wanting even more info (for example, i rewrote `Regexp.match?` to see how the flame changed, then restored that and rewrote a different method and made another one...)

Right. Then maybe a suffix like you added, which just defaults to `RUBY_ENGINE`?

rwstauner commented:
I switched to `stackprof` and simplified it a bit.  This makes the behavior with CRuby pretty close to the way the `profile` command works normally with TruffleRuby.
How do you feel about this?

eregon reviewed (commented):
This looks much simpler, thanks!

BTW I was curious if there is a way to profile requiring RubyGems and there is, via:
```
jt -u ruby profile --disable-gems -I ~/.rubies/ruby-3.1.3/lib/ruby/gems/3.1.0/gems/stackprof-0.2.25/lib -e 'require "rubygems"'
```

rwstauner commented on tool/jt.rb:
```diff
@@ -2094,19 +2094,42 @@ def profile(*args)
+    if truffleruby?
+      app_open svg_filename
+    else
+      File.open(flame_file, 'w') do |f|
+        run_ruby(env, '-S', 'stackprof', '--d3-flamegraph', flame_file_dump, {:out => f})
+      end
```
sadly `--out` is only available for the `run` command, not the post-processing:
https://github.com/tmm1/stackprof/blob/master/bin/stackprof#L10-L18

eregon replied:
Ah yes, I misread the help output. Then we can just use `out: flame_file` to simplify, I'll push a commit for that.
